### ISSUE TITLE
createPortal: support option to stop propagation of events in React tree

### ISSUE THREAD

**Issue by kib357** [facebook/react#11387]
**Do you want to request a *feature* or report a *bug*?**
Feature, but also a bug cause new API breaks old `unstable_rendersubtreeintocontainer`

**What is the current behavior?**
We cannot stop all events propagation from portal to its React tree ancestors. Our layers mechanism with modals/popovers completely broken. For example, we have a dropdown button. When we click on it, click opens popover. We also want to close this popover when clicking on same button. With createPortal, click inside popover fires click on button, and it's closing. We can use stopPropagation in this simple case. But we have tons of such cases, and we need use stopPropagation for all of them. Also, we cannot stop all events.

**What is the expected behavior?**
createPortal should have an option to stop synthetic events propagation through React tree without manually stopping every event. What do you think?

**Comment by kib357** [comment]
Also, propagation of mouseOver/Leave looks completely unexpected.
![image](https://user-images.githubusercontent.com/347668/32112901-43f72342-bb47-11e7-8f2f-0e1b2e7e1646.png)

**Comment by gaearon** [comment]
Can you move portal outside of the button?

e.g.

```js
return [
  <div key="main">
    <p>Hello! This is first step.</p>
    <Button key="button" />
  </div>,
  <Portal key="portal" />
];
```

Then it won't bubble through the button.

**Comment by kib357** [comment]
It was my first thought, but!) Imagine, that we have mouseEnter handler in such component container:

![image](https://user-images.githubusercontent.com/347668/32129253-ced34e14-bb8d-11e7-82f1-f9f857fa50e6.png)

With `unstable_rendersubtreeintocontainer` i need nothing to do with events in `ButtonWithPopover` component – mouseEnter simply works when mouse really enters `div` and button DOM element, and not fired when mouse is over popover. With portal, event fires when mouse over popover – and actually NOT over `div` in this moment. So, i need to stop propagation. If i do it in `ButtonWithPopover` component, i will break event firing when mouse is over button. If i do it in popover and i'm using some common popover component for this application, i also can break logic in other app parts.

I really don't understand purpose of bubbling through React tree. If i need events from portal component – i simply can pass handlers through props. We were do it with `unstable_rendersubtreeintocontainer` and it worked perfect.

If i will open a modal window from some button deep in react tree, i will receive really unexpected firing of events under modal. `stopPropagation` will also stop propagation in DOM, and i will not get events that i really expect to be fired(

**Comment by craigkovatch** [comment]
@gaearon I would suggest this is more of a bug than a feature request. We have a number of new bugs caused by mouse events bubbling up through portals (where we were previously using `unstable_rendersubtreeintocontainer`). Is there a way to workaround this before this is addressed in a future release?

**Comment by jquense** [comment]
I think it's being called a feature request, because the current bubble behavior of portals is both expected and intended. The goal is that subtree act like real child of their parents.

What would be helpful is additional use cases or situations (like the ones you're seeing) that you don't feel are served by the current implementation, or are difficult to workaround

**Comment by craigkovatch** [comment]
I understand that this behavior is intended, but I think it's a significant bug that it's not disable-able.

**Comment by neytema** [comment]
In my mind library working with DOM should preserve DOM implementation behavior not break it.

For example:

```jsx
class Container extends React.Component {
  shouldComponentUpdate = () => false;
  render = () => (
    <div
      ref={this.props.containerRef}
      // Event propagation on this element not working
      onMouseEnter={() => { console.log('handle mouse enter'); }}
      onClick={() => { console.log('handle click'); }}
    />
  )
}

class Root extends React.PureComponent {
  state = { container: null };
  handleContainer = (container) => { this.setState({ container }); }

  render = () => (
    <div>
      <div
        // Event propagation on this element not working also
        onMouseEnter={() => { console.log('handle mouse enter'); }}
        onClick={() => { console.log('handle click'); }}
      >
        <Container containerRef={this.handleContainer} />
      </div>
      {this.state.container && ReactDOM.createPortal(
        <div>Portal</div>,
        this.state.container
      )}
    </div>
  );
}
```

When I work with DOM, I expect to receive events like DOM implementation does it. In my example events are propagated through **Portal**, working around it's DOM parents, and this can be considered as a **bug**.

**Comment by jquense** [comment]
Folks thanks, for the discussion but I don't think it's all that helpful to argue whether something is a bug or not. Instead i'd be more productive to discuss the use cases and examples that are not met by the current behavior, so we can better understand if the current way is the best way for the future.

In general we want the API to handle a diverse set of use-cases while hopefully not overly limiting others. I can't speak for the core team, but I'd imagine that making it configurable is not a likely solution. Generally React leans for a consistent api over configurable ones. 

 I also understand that this behavior is not how the DOM works, but i don't think that's in itself a good reason to say it shouldn't be that way. Lots of react-dom's behavior is different from how the DOM works, may events are already different from the native version. `onChange` for instance is completely unlike the native change event, and all react events bubble regardless of type, unlike the DOM.

**Comment by craigkovatch** [comment]
> Instead i'd be more productive to discuss the use cases and examples that are not met by the current behavior

Here's two examples that are broken for us in our migration to React 16.

First, we have a draggable dialog which is launched by a button. I attempted to add a "filtering" element on our Portal use which called StopPropagation on any mouse* an key* events. However, we rely on being able to bind a mousemove event to the document in order to implement the dragging functionality -- this is common because if the user moves the mouse at any significant rate, the cursor leaves the bounds of the dialog and you need to be able to capture the mouse movement at a higher level. With Portals, the mouse and key events are bubbling up from inside the dialog to the button that launched it, causing it to display different visual effects and even dismiss the dialog. I don't think it's realistic to expect every component that will be launched via a Portal to bind 10-20 event handlers to stop this event propagation.

Second, we have a popup context menu which can be launched by either a primary- or secondary-mouse click. One of the internal consumers of our library has mouse handlers attached to the element that launches this menu, and of course the menu also has click handlers for handling item selection. The menu is now reappearing on every click as the mousedown/mousedown events are bubbling back up to the button that launches the menu.

> I can't speak for the core team, but I'd imagine that making it configurable is not a likely solution. Generally React leans for a consistent api over configurable ones.

I implore you (and the team) to reconsider this position in this particular case. I think event bubbling will be interesting for certain use cases (although I can't think of any offhand). But I think it will be crippling in others, and it introduces significant inconsistency in the API. While `unstable_rendersubtreeintocontainer` was never super-supported, it was what everyone used to render outside of the immediate tree, and it didn't work this way. It was officially deprecated in favor of Portals, but Portals break the functionality in this critical way, and there doesn't seem to be an easy workaround. I think this can be fairly described as quite inconsistent.

> I also understand that this behavior is not how the DOM works, but i don't think that's in itself a good reason to say it shouldn't be that way.

I understand where you're coming from here, but I think in this case (a) it's a fundamental behavior which (b) currently has no workaround, so I think "the DOM doesn't work this way" is a strong argument, if not a completely convincing one.

**Comment by craigkovatch** [comment]
And to be clear: my request that this be considered a bug is mostly so that it gets prioritized for a fix sooner rather than later.

**Comment by sebmarkbage** [comment]
My mental model of a Portal is that it behaves as if it is on the same place in the tree, but avoids problems such as "overflow: hidden" and avoids scrolling for drawing/layout purposes.

There are many similar "popup" solutions that happen inline without a Portal. E.g. a button that expands a box right next to it.

Take as an example the "Pick your reaction" dialog here on GitHub. That is implemented as a div right next to the button. That works fine now. However, if it wants to have a different z-index, or be lifted out of an `overflow: scroll` area that contains these comments, then it will need to change DOM position. That change is not safe to do unless other things like event bubbling is also preserved.

Both styles of "popups" are legit. So how would you solve the same problem when the component is inline in the layout as opposed to floating outside of it?

**Comment by methyl** [comment]
The workaround that worked for me is calling `stopPropagation` directly under portal rendering:
```
return createPortal(
      <div onClick={e => e.stopPropagation()}>{this.props.children}</div>,
      this.el
    )
```

That works great for me since I have single abstraction component that uses portals, otherwise you will need to fix up all your `createPortal` calls.

**Comment by craigkovatch** [comment]
> Both styles of "popups" or "pop outs" are legit. So how would you solve the same problem when the component is inline in the layout as opposed to floating outside of it?

I'm not sure this question makes sense. If I had this problem inlining the component, I wouldn't inline it.

**Comment by jquense** [comment]
I think some of problem here is the some use cases of `renderSubtreeIntoContainer` are being ported to `createPortal` when the tools are doing conceptually different things. The concept of Portal was being overloaded I think.

I agree that in the Modal dialog case, you almost never want the modal to act like a child of the button that opened it. I think tho it's a mistake to say that the portal implementation is therefore wrong, instead of saying that `createPortal` in the button is not the right tool for this. In this case the Modal is not a child of the trigger, and shouldn't be rendered as if it were. One possible solution is to keep using `renderSubtreeIntoContainer`, another user-land option is to have a `ModalProvider` near the app root that handles rendering modals, and passes down (via context) a method to render an arbitrary modal element need to the root

**Comment by craigkovatch** [comment]
`renderSubtreeIntoContainer` can't be called from inside of `render` or lifecycle methods in React 16. Portals are the official recommendation: https://reactjs.org/blog/2017/09/26/react-v16.0.html#breaking-changes

I agree that the concept of Portals might have ended up overloaded. Not sure I love the solution of a global component and context for it, though. It seems like this could be easily solved by a flag on createPortal specifying whether events should bubble. (Or perhaps a prop with a whitelist of events that *should* bubble?)

**Comment by kib357** [comment]
I will try to clarify our use-case of the portals and why we would love to see an option for stopping events propagation. In ManyChat app, we are using portals to create a 'layers'. We have the layer system for the whole app that used by several types of components: popovers, dropdowns, menus, modals. Every layer can expose a new layer, e.g. button on a second level of menu can trigger the modal window where the other button can open the popover. In most cases layer is the new branch of UX that solves it's own task. And when new layer are open, the user should interact with this new layer, not with others in bottom. So, for this system, we've created a common component for rendering to layer:
```JavaScript
class RenderToLayer extends Component {
  ...
  stop = e => e.stopPropagation()

  render() {
    const { open, layerClassName, useLayerForClickAway, render: renderLayer } = this.props

    if (!open) { return null }

    return createPortal(
      <div
        ref={this.handleLayer}
        style={useLayerForClickAway ? clickAwayStyle : null}
        className={layerClassName}
        onClick={this.stop}
        onContextMenu={this.stop}
        onDoubleClick={this.stop}
        onDrag={this.stop}
        onDragEnd={this.stop}
        onDragEnter={this.stop}
        onDragExit={this.stop}
        onDragLeave={this.stop}
        onDragOver={this.stop}
        onDragStart={this.stop}
        onDrop={this.stop}
        onMouseDown={this.stop}
        onMouseEnter={this.stop}
        onMouseLeave={this.stop}
        onMouseMove={this.stop}
        onMouseOver={this.stop}
        onMouseOut={this.stop}
        onMouseUp={this.stop}

        onKeyDown={this.stop}
        onKeyPress={this.stop}
        onKeyUp={this.stop}

        onFocus={this.stop}
        onBlur={this.stop}

        onChange={this.stop}
        onInput={this.stop}
        onInvalid={this.stop}
        onSubmit={this.stop}
      >
        {renderLayer()}
      </div>, document.body)
  }
  ...
}
```
This component stops propagation for all event types from React docs, and it allowed us to update to React 16.

**Comment by jacobp100** [comment]
Does this need to be tied to portals? Rather than sandboxing portals, what if there was just a (for example) `<React.Sandbox>...</React.Sandbox>`?

**Comment by craigkovatch** [comment]
Even that seems needlessly complex to me. Why not simply add an optional boolean flag to createPortal allowing the bubbling behavior to be blocked?

**Comment by craigkovatch** [comment]
@gaearon this is a pretty unfortunate situation for a certain slice of us -- could you or someone dear to you have a look at this? :)

**Comment by jquense** [comment]
I'd add that my current thinking is that both use cases should be supported. There are really use cases where you need context to flow from the current parent to the subtree but to not have that subtree act as a logical child in terms of the DOM. Complex modals are the best example, you just almost never want the events from a form in a modal window to propagate up to the trigger button, but almost certainly need the context passed through (i18n, themes, etc)

I will say that that usecase _could_ be mostly solved with a ModalProvider closer to the app root that renders via `createPortal` high enough that event propagation doesn't affect anything, but that starts to feel like a workaround as opposed to a well designed architecture. It also makes library provided modals more annoying for users since they are no longer self contained.

**Comment by jquense** [comment]
i would add tho in terms of API i don't think `createPortal` should do both, the modal case really wants to just use `ReactDOM.render` (old skool) because it's pretty close to a distinct tree _except_ that context propagation is often needed

**Comment by craigkovatch** [comment]
We just had to fix an extremely difficult-to-diagnose bug in our outer application's focus management code as a result of using the workaround that @kib357 posted.

Specifically: calling stopPropagation on the synthetic focus event to prevent it from bubbling out of the portal causes stopPropagation to also be called on the native focus event in React's captured handler on #document, which meant it did not make it to another captured handler on `<body>`. We fixed by moving our handler up to #document, but we had specifically avoided doing that in the past so as not to step on React's toes.

The new bubbling behavior in Portals really feels like the minority case to me. Be that opinion or truth, could we please get some traction on this issue? Maybe @gaearon? It's four months old and causing real pain. I think this could fairly be described as a bug given that it's a breaking API change in React 16 with no completely-safe workaround.

**Comment by sebmarkbage** [comment]
@craigkovatch I'm still curious how you would solve my inline example. Let's say the popup is pushing the size of the box down. Inlining something is important since it is pushing something down in the layout given its size. It can't just hover over.

You could potentially measure the popover, insert a blank placeholder with the same size and try to align it on top but that's not what people do.

So if your popover need to expand the content in place, like right next to the button, how would you solve it? I suspect that the pattern that works there, will work in both cases and we should just recommend the one pattern.

**Comment by sebmarkbage** [comment]
I think in general this is the pattern that works in both scenarios:

```js
class Foo extends React.Component {
  state = {
    highlight: false,
    showFlyout: false,
  };

  mouseEnter() {
    this.setState({ highlight: true });
  }

  mouseLeave() {
    this.setState({ highlight: false });
  }

  showFlyout() {
    this.setState({ showFlyout: true });
  }

  render() {
    return <>
      <div onMouseEnter={this.mouseEnter} onMouseLeave={this.mouseLeave}>
        Hello
        <Button key="button" onClick={this.showFlyout} />
      </div>
      {this.state.showFlyout ? <Flyout /> : null}
    </>;
  }
}
```

If it is a portal, then it works and it doesn't get any mouse over events when hovering over the portal. But more importantly, it also works if it is NOT a portal, and it needs to be an inline flyout.

So what is it about this pattern that doesn't work for your use case?

**Comment by craigkovatch** [comment]
@sebmarkbage we are using Portals in a completely different fashion, rendering into a container mounted as the final child of `<body>` which is then positioned, sometimes with a z-index. The React documentation suggests this is closer to the design intention; i.e. rendering into a totally different place in the DOM. It doesn't seem to me that our use cases are similar enough for discussion to belong on this thread. But if you want to brainstorm/troubleshoot together, I'd be more than happy to discuss further in another forum.

**Comment by sebmarkbage** [comment]
No my use case is *both*. Sometimes one and sometimes the other. That's why it is relevant.

The `<Flyout />` can choose to render into the final child of body or not but as long as you hoist the portal itself out to a sibling of the hovered component rather than a child of it, your scenario works.

**Comment by sebmarkbage** [comment]
I think there's a plausible scenario where that is inconvenient and you want a way to teleport things from deeply nested components but in that scenario you're probably fine with the context being the context from the intermediate point. But I think of those as two separate issues.

Maybe we need a slots API for that.

```js
class Foo extends React.Component {
  state = {
    showFlyout: false,
  };

  showFlyout() {
    this.setState({ showFlyout: true });
  }

  hideFlyout() {
    this.setState({ showFlyout: false });
  }

  render() {
    return <>
      Hello
      <Button onClick={this.showFlyout} />
      <SlotContent name="flyout">{this.state.showFlyout ? <Flyout onHide={this.hideFlyout} /> : null}</Slot>
    </>;
  }
}

class Bar extends React.Component {
  state = {
    highlight: false,
  };

  mouseEnter() {
    this.setState({ highlight: true });
  }

  mouseLeave() {
    this.setState({ highlight: false });
  }

  render() {
    return <>
      <div onMouseEnter={this.mouseEnter} onMouseLeave={this.mouseLeave} className={this.state.highlight ? 'highlight' : null}>
        <SomeContext>
          <DeepComponent />
        </SomeContext>
      </div>
      <Slot name="flyout" />
    </>;
  }
}
```

The portal would then get the context of Bar, not DeepComponent. Context and event bubbling then still share the same tree path.

**Comment by jquense** [comment]
@sebmarkbage the modal case usually does require context from the point it's rendered. It's slightly unique of a case I think, the component is a logical child of the thing that rendered it but _not_ a structural one (for lack of a better word), e.g. You usually want things like form context (relay, formik, redux form, whatever) but not DOM events to pass through. One also ends up rendering such modals fairly deep in trees, next to their triggers, so they stay componenty and reusable, more so than because they belong there structurally

I think this case is generally different to the  flyout/dropdown case createPortal serves. Tbc I think the bubbling behavior of portals is good, but not for modals. I also think this could handled with Context and some sort of ModalProvider reasonably well, but that's kinda annoying especially for libraries.

**Comment by craigkovatch** [comment]
> as long as you hoist the portal itself out to a sibling of the hovered component rather than a child of it, your scenario works.

Not sure I follow. There's still the problem of e.g. keyDown events bubbling through an unexpected DOM tree.

**Comment by sebmarkbage** [comment]
@jquense Note that in my example the slot is still within the Bar component so it would get its context from the form in something like <Form><Bar /></Form>.

Even if the portal rendered into the document body.

So it’s like two indirections (portalings): deep -> sibling of Bar -> document body.

So the context of the portal is still the context of the form, and so is the event bubbling chain, but neither is in the context of the hovered thing.

**Comment by jquense** [comment]
Yes sorry missed that 😳 If I'm reading it right tho, you'd still have the bubbling at `<Slot>` up though?  That's definitely better, though I do think that in the Modal dialog case one probably doesn't want _any_ bubbling. Like thinking in terms of a screen reader, you want everything outside the modal to be invert, while it's up. I don't know, I think for that case the bubbling is a gotcha, no one would expect a click inside a dialog to bubble through anywhere.

**Comment by jquense** [comment]
Maybe the problem here isn't portals, but there isn't a good way to share Context across trees? A part from the context thing `ReactDOM.render` is really fine for modals, and maybe more "correct" way of thinking about it anyway...

**Comment by sebmarkbage** [comment]
My thinking here is that there is *some* bubbling because it still goes from the modal to the div to the body to the document to the window. And conceptually beyond the frame out to the containing window and so on.

That's not theoretical in something like ART or GL rendered content (and to some extent React Native) where there might not be an existing backing tree to get those semantics from. So there needs to be a way to say that *this* is where it bubbles.

In some apps there are modals in modals. E.g. at FB there is chat window that might be above a modal or a modal might be part of the chat window. So even a modal has some context as to where in the tree it belongs. It is never completely standalone.

That's not to say we can't have two different semantics for event bubbling and context. That are both explicit about this and you can portal one without the other etc.

Having the guarantee that they both follow the same path is really powerful though since it means that event bubbling can be fully implemented for user space events the same as in the browser.

For example this happens with various Redux contexts today. Imagine `this.context.dispatch("Hover")` is a user space event bubbling. We could even implement React events as part of context. It seams reasonable to think that I can use this the same way, and in every way right now, you can. I think if we did fork these two contexts, we'd probably end up with another user space context API that follows the DOM structure in parallel with normal context - if they are truly so different.

So that's kind of why I'm pushing against it a bit to see if the slots thing might be sufficient, since a) you need to be explicit about which context bubbling happens anyway. b) it can avoid forking the world and having two entire context systems.

**Comment by kib357** [comment]
> Specifically: calling stopPropagation on the synthetic focus event to prevent it from bubbling out of the portal causes stopPropagation to also be called on the native focus event in React's captured handler on #document, which meant it did not make it to another captured handler on <body>. We fixed by moving our handler up to #document, but we had specifically avoided doing that in the past so as not to step on React's toes.

@craigkovatch , did you used `onFocusCapture` event on document? In my workaround captured events shouldn't be stopped. Can you provide more detailed example of how it was and what you have done to resolve your issue?
Also, i think my code has an issue with stopping `blur` event – it shouldn't be stopped. So, i'll investigate this question deeper and will try to found a more reliable solution.

**Comment by craigkovatch** [comment]
@kib357 I'm not suggesting there's an issue in your workaround, I think there's a separate bug in React there (i.e. shouldn't cancel propagation of focus events in capture phase when calling stopPropagation on synthetic focus events in bubbling phase).

The code in question uses a native capture event listener, i.e. `document.body.addEventListener('focus', handler, true)'

**Comment by kib357** [comment]
@craigkovatch sounds interesting, given the fact that you used captured handler. However, i have not any thoughts why this happens.

So, guys, we have a two different scenarios for using portal rendering:

1. To prevent CSS issues like overflow:hidden and etc in simple widgets, like a dropdown buttons or one-level menus
2. To create a new UX layer for more powerful cases like:
- modals
- nested menus
- popovers-with-forms-with-dropdowns-... – all cases, when layers are combined

I think current `createPortal` API satisfies only the first scenario. Suggestion to use a new React.render for second is unusable – it's very poor to create a separate app with all it providers for every layer.
What additional info we can provide to help resolve this issue?
What disadvantages of suggested param in the `createPortal` API?

**Comment by webdesserts** [comment]
@sebmarkbage My immediate question with the slots API is: would I be able to insert multiple `SlotContents` into one `Slot` at the same time? It's not uncommon in our interface to have multiple "popups" or "modals" open simultaneously. In my perfect world a `Popup` API would look something like this:

```tsx
import { PopupSlot } from './popups'

let root = (
  <div>
    <SomeProvider>
      <App />
      <PopupSlot />
    </SomeProvider>
  </div>
)

ReactDOM.render(root, document.querySelector('#root'))

```

```tsx
// some dark corner of our app

import { Popup } from './popups'

export function SoManyPopups () {
  return <>
    <Popup>My Entire</Popup>
    <Popup>Interface</Popup>
    <Popup>Is Popups</Popup>
  </>
}
```

**Comment by craigkovatch** [comment]
TL;DR: there are major issues even with the workaround discussed in here. The lack of response from the React team is extremely frustrating. This is a breaking API change and should be treated as a significant bug, not a feature request.

We have a new issue with this that I have been completely unable to find a workaround for. Using the "event trap" approach suggested above, only React Synthetic events are blocked from bubbling out of the portal. The native events still bubble, and since our React code is hosted inside of a mostly-jQuery application, the global jQuery keyDown handler on `<body>` still gets the event.

I attempted to add an event.stopPropagation listener to the native container element inside the Portal via a ref like this, but this completely neuters all Synthetic events within the portal -- although I am completely flabbergasted as to why (why should bubbling-out native events interfere with React Synthetic events?)

```
const allTheEvents: string[] = 'click contextmenu doubleclick drag dragend dragenter dragexit dragleave dragover dragstart drop mousedown mouseenter mouseleave mousemove mouseover mouseout mouseup keydown keypress keyup focus blur change input invalid submit'.split(' ');
const stop = (e: React.SyntheticEvent<HTMLElement>): void => { e.stopPropagation(); };
const nativeStop = (e: Event): void => e.stopPropagation();
const handleRef = (ref: HTMLDivElement | null): void => {
  if (!ref) { return; }
  allTheEvents.forEach(eventName => ref.addEventListener(eventName, nativeStop));
};


/** Prevents https://reactjs.org/docs/portals.html#event-bubbling-through-portals */
export function PortalEventTrap(children: React.ReactNode): JSX.Element {
  return <div
      onClick={stop}
      ...

      ref={handleRef}
    >
      {children}
    </div>;
```

**Comment by Kovensky** [comment]
That depends on the order that ReactDOM and JQuery are initialized. If JQuery initializes first, JQuery's top level event handlers will be installed first.

Both ReactDOM and JQuery prefer to only have a single top level listener that then simulates bubbling internally, unless there's some event that the browser won't bubble such as `scroll`.

**Comment by craigkovatch** [comment]
@Kovensky my understanding was that jQuery did not do "synthetic bubbling" the way React does, and thus does not have a single top-level listener. My DOM inspector doesn't reveal one, either. Would love to see what you're referencing if I'm mistaken.

**Comment by Kovensky** [comment]
That will be the case for delegated events. For example, `$(document.body).on('click', '.my-selector', e => e.stopPropagation())`.

**Comment by sebmarkbage** [comment]
Look, this can be solved in React, if someone just convinces me that this can’t be solved with my proposed design above which requires some restructuring of your code. But I haven’t seen any reason that can’t be done other than just trying to find a quick fix workaround.

**Comment by craigkovatch** [comment]
@sebmarkbage your proposal only solves the case of the events propagating to the immediate owner. What about the rest of the tree?

**Comment by jquense** [comment]
Here is a use case I think can't be solved well with Slots or createPortal

```jsx
<Form defaultValue={fromValue}>
   <more-fancy-markup />
   <div>
     <Field name="faz"/>
     <ComplexFieldModal>
       <Field name="foo.bar"/>
       <Field name="foo.baz"/>
     </ComplexFieldModal>
  </div>
</Form>
```

And here is a gif with a similar but slightly different setup, where i'm using createPortal for a responsive site, to move a form field to the app toolbar (much higher up the tree). In this case as well I really don't want events bubbling back to the page content, but i definitely want the Form context to go with it. My implementation btw is some Slot-esque thing using context...

![large gif 640x320](https://user-images.githubusercontent.com/339286/39369169-8d7d4c5c-4a09-11e8-8d0d-830d2a14f07e.gif)

**Comment by JasonGore** [comment]
@sebmarkbage `unstable_renderSubtreeIntoContainer` allowed direct access to the top of the hierarchy regardless of a component's position, either within the hierarchy or as part of a separate packaged framework. 

Comparatively, I see a few problems with the Slot solution:
- The solution assumes you have access to the position of the hierarchy where it is "ok" to bubble events. This is definitely not the case for components and component frameworks.
- Assumes it is "ok" to bubble events at any other level of the hierarchy.
- Events will still bubble up from Slot's position. (as @craigkovatch mentioned)

**Comment by slorber** [comment]
I also have a usecase (probably similar to ones already mentionned).

I have a surface where users can select things with his mouse with a "lasso". This is basically 100% width/height and is at the root of my app and use `onMouseDown` event.  In this surface there are also buttons that open portals like modals and dropdowns. A mouseDown event inside the portal is actually intercepted by the lasso selection component at the root of the app.

I see many to fix the problem:

- render the portal one step above the root lasso component, but this is not very convenient and would probably need to resort to a context-based lib like react-gateway? (or maybe the slot system mentionned).
- stop propagation manually inside the portal root, but could lead to unwanted side-effects mentionned above
- ability to stop propagation in React portals (+1 btw)
- filter out events when they come from a portal 



For now my solution is to filter out the events.

```js
const appRootNode = document.getElementById('root');

const isInPortal = element => isNodeInParent(element, appRootNode);


    handleMouseDown = e => {
      if (!isInPortal(e.target)) {
        return;
      }
      ...
    };
```

This will clearly not be the best solution for all of us and will not be very nice if you have nested portals, but for my current usecase (which is the only one currently) it works. I don't want to add a new context lib or do a complex refactor to solve this. Just wanted to share my solution.

**Comment by JasonGore** [comment]
I've been able to accomplish blocking event bubbling as noted elsewhere in this thread. 

But another seemingly thornier issue I'm running into is the `onMouseEnter` SyntheticEvent, which does not bubble up. Rather it traverses from the common parent of the `from` component to the `to` component as described [here](https://reactjs.org/docs/events.html#mouse-events). This means that if the mouse pointer enters from outside of the browser window, every `onMouseEnter` handler from the top of the DOM down to the component in createPortal will be triggered in that order, causing all sorts of events to fire that never did with `unstable_renderSubtreeIntoContainer`. Since `onMouseEnter` doesn't bubble up, it can't be blocked at the Portal level. (This did not seem to an issue with `unstable_renderSubtreeIntoContainer` as the `onMouseEnter` event did not respect virtual hierarchy and did not sequence through the body content, rather descended directly into the subtree.)

If anyone has any ideas on how to prevent `onMouseEnter` events from propagating from the top of the DOM hierarchy or divert directly into the portal subtree, please let me know.

**Comment by hegelstad** [comment]
@JasonGore I've also noticed this behavior.

For instance.

I have a context menu that is rendered when a div triggers onMouseOver, then I open a Modal with createPortal by clicking one of the items in the menu. When I bring the mouse out of the browser window, the onMouseLeave event propagates up to the context menu, closing the context menu (and thus the Modal)...

**Comment by jnsandrew** [comment]
Had the same issue where I had a list item that I wanted the entirety to be clickable (as a link), but wanted a delete button on the side which would open a modal to confirm. My only solution was to prevent bubbling on the modal div like so:

```javascript
// components/Modal.js

onClick(e) {
    e.stopPropagation();
}

return createPortal(
            <div onClick={this.onClick} ...
                ...
```

It'll prevent bubbling on every modal yes, but I have no case where I would want that to happen yet so it works for me.

**Comment by AlastairTaft** [comment]
Just hit this. It seems awkward to me that React would behave in its own way that is different to DOM event bubbling.

**Comment by danielhusar** [comment]
+1 to this. We are using `React.createPortal` to render inside the iframe, (for both styles and events isolation) and not be able to prevent events to bubble out of the box is a bummer.

**Comment by jcrben** [comment]
Looks like this is the 12th most-thumbed up issue in React's backlog. At least the docs are open about it https://reactjs.org/docs/portals.html#event-bubbling-through-portals - altho they don't mention the downsides or workarounds and instead note:
> Catching an event bubbling up from a portal in a parent component allows the development of more flexible abstractions that are not inherently reliant on portals. For example, if you render a <Modal /> component, the parent can capture its events regardless of whether it’s implemented using portals.

The docs should at least explain that this can cause problems and suggest workarounds. In my case, it's a pretty straightforward use case, using https://github.com/reactjs/react-modal: I have buttons which open things like dropdowns, and inside these are buttons which create modals. Clicking the modal bubbles up to top button, causing it to do unwanted  stuff. The modals are encapsulated into a cohesive component and pulling out the portaled part breaks that encapsulation, creating a leaky abstraction. One workaround might flipping a flag to disable these buttons while the modal is open.

I'm not sure how often it's really helpful bubbling and capture are in general - they certainly have a storied history, but I'd rather pass a callback or propagate a more specific event (e.g., redux action) than bubble up or capture down, since such things probably through a bunch of unnecessary intermediaries. There's articles like https://css-tricks.com/dangers-stopping-event-propagation/ and I work on apps which do depend on propagation to the body, mostly to close things when clicking "outside", but I'd rather put an invisible overlay over everything and close on clicking that. Of course, I couldn't use React's Portal to create such an invisible overlay...

**Comment by craigkovatch** [comment]
There's also a maintenance nightmare here -- as new events get added to the DOM, any portals "sealed" with the above-discussed technique will "leak" those new events until maintainers can go add them to the blacklist.

There's a major design issue here that needs to be addressed. An ability to opt-in or opt-out of cross-Portal bubbling still seems the best API option to me. Not sure about implementation difficulty therein, but we're still getting production bugs around this in Tableau, over a year later.

**Comment by Arseniy-II** [comment]
Spend 2 hours trying to find out why my form from modal was submitting another form.
Finally I figured it out thanks to that issue!

I really struggle to see when `onSubmit` propagation may be necessary. Most likely it will always be more like a bug rather then a feature.

**Comment by stlegion** [comment]
+1 to this also. We are using [draftjs](https://draftjs.org/) heavilly with clickable text showing modals. And all events on modal like focus, select, change, keypress etc. just exploding draftjs with errors.

**Comment by dcporter** [comment]
IMO, the event-proxying behavior is fundamentally broken (and is causing me bugs as well), but I recognize that this is controversial. This thread pretty strongly suggests that there's a need for a portal which wormholes *context* but not *events*. Does the core team agree? Either way, what's the next step here?

**Comment by antonychopin** [comment]
I cannot really realize, why propagating events from portal is intended behaviour? This is completely stands against the main idea of propagation. I thought that portals were created exactly to avoid this kind of things (like manual nesting, events propagation, etc.).

I can confirm, that if you put the portal near the elements tree, then it won't propagate events:

```tsx
class SomeComponent extends React.Component<any, any> {
  render() {
    return <>
      <div className="some-tree">
        // Portal here will bubble events
      </div>
      // Portal here won't bubble events, just checked
    </>
  }
}
```
But this is a kind of inconvenience in case of a huge components, I think. Also this is inconvenient in case of dynamical rendering (e.g. when mapping an array or something).
+1 for this feature request

**Comment by jamiewinder** [comment]
In DOM, events bubble up the DOM tree. In React, events bubble up the component tree.

I rely on this behaviour quite a bit, an example of which is popouts which may be nested; they're all portals to avoid issues with `overflow: hidden`, but in order to get the popout to behave correctly I need to detect external clicks to the popout component (which is different to detecting clicks outside of the rendered DOM elements). There may be better examples.

**Comment by craigkovatch** [comment]
I think the robust discussion here has made it clear that there are good reasons to have *both* behaviors. Because `createPortal` renders a React component inside of a "plain DOM" container node, I don't think it would be workable for React's Synthetic Events to propagate out of a Portal up the plain-old-DOM tree.

Since Portals have been out for a long time now, it's probably too late to change the default behavior to "don't propagate past the portal boundary".

Based on all the discussion so far, my simplest proposal is then (still): add an optional flag to createPortal that prevents *any* event propagation past the portal boundary.

Something more robust might be the ability to provide a whitelist of events that should be allowed to "poke through" the boundary, while stopping the rest.

@gaearon Are we at the point that the React team could actually take this on? This is a top-10 issue but we haven't heard anything from yous on this in quite a while.

**Comment by dgreensp** [comment]
I want to add my support to this, and disagree with @sebmarkbage's comments from last year arguing that portaling both React context and DOM event bubbling makes more conceptual sense than portaling just context.

The ability to portal context from one place in the DOM to another is useful for implementing all manner of overlays such as tooltips, dropdowns, hover cards, and dialogs, where the content of the overlay is described by, and rendering in the context of, the trigger.  Since context is a React concept, this mechanism solves a React problem.  On the other hand, the ability to portal DOM event bubbling from one place in the DOM to another is a fancy trick that lets you pretend the DOM structure is other than what you explicitly set it up to be.  This solves a problem with using DOM event bubbling for delegation, when you want to delegate to a different part of the DOM.  Probably you should be using callbacks (or context) anyway, if you have React, rather than relying on DOM events bubbling from inside to outside of the overlay.  As others have pointed out, rarely do you want to "reach in" and handle an event happening inside the overlay, intentionally or unintentionally.

DOM event bubbling primarily solves a problem of matching DOM events to DOM targets.  Every click is actually a click on a whole set of nested elements.  It's not best thought of as a high-level delegation mechanism, IMO, and using DOM events to delegate across React component boundaries is not great encapsulation, unless the components are small private helper components used to render predictable bits of DOM.

**Comment by Fi1osof** [comment]
event.target === event.currentTarget helps me solve this problem. But this is realy headache.

**Comment by mogelbrod** [comment]
This bit me today while trying to migrate a popover component using `unstable_renderSubtreeIntoContainer` to use `createPortal`.  The component in question contains draggable elements and is rendered as a descendant of another draggable element. This means both the parent and popover elements contain mouse & touch event handlers, which both started firing when interacting with the portal'ed popover.

Since `unstable_renderSubtreeIntoContainer` [is being deprecated](https://github.com/facebook/react/pull/12730)(?) an alternative solution is necessary - none of the workarounds presented above appears to be viable long term solutions.

**Comment by Mobilunity-Com** [comment]
Hey! Thank for all this suggestions guys! 
It helped me to fix one of my troubles. 
Would you like to read great and informative article about the importance and abilities of [React team](https://mobilunity.com/blog/a-dedicated-reactjs-developer/)? I guess it will be useful for everybody who is interested in development. Good luck!

**Comment by dinofx** [comment]
IMO it's more often you want a portal to give you access to context, but not bubble up events. Back when we used Angular 1.x, we wrote our own popup service that would take `$scope` and a template string, and compile/render that template and append it to the body.  We implemented all of our application's popups/modals/dropdowns with that service, and not once did we miss the lack of event bubbling.

**Comment by mogelbrod** [comment]
The [`stopPropagation()` workaround](https://github.com/facebook/react/issues/11387#issuecomment-355395803) appears to prevent native event listeners on `window` from triggering (in our case added by `react-dnd-html5-backend`).

Here's a minimal repro of the issue: https://codepen.io/mogel/pen/xxKRPbQ

If there's no plan of providing a way to avoid the synthetic bubbling across portals, perhaps someone has a workaround which doesn't break native event bubbling?

**Comment by dcporter** [comment]
> The stopPropagation() workaround appears to prevent native event listeners on window from triggering

Correct. :(

> If there's no plan of providing a way to avoid the synthetic bubbling across portals

Despite the core team's silence, I, and many others on this thread, _really hope_ that there are such plans.

> perhaps someone has a workaround which doesn't break native event bubbling?

My team's workaround has been to ban portals entirely because of this glaring issue. We present panes with a hook into a container that lives within the app’s other contexts, so you get root level contexts for free; any others we pass across manually. Not great, but better than pointless whack-a-mole event handlers.

**Comment by mogelbrod** [comment]
It's been 17 months since [the last response](https://github.com/facebook/react/issues/11387#issuecomment-371989366) by anyone from the core team. Maybe a ping could get this issue some attention :) @sebmarkbage or @gaearon

> My team's workaround has been to ban portals entirely because of this glaring issue. We present panes with a hook into a container that lives within the app’s other contexts, so you get root level contexts for free; any others we pass across manually. Not great, but better than pointless whack-a-mole event handlers.

I can't think of any generic approaches to passing context down into the "fake portal" via props without going back to relying on cascading props :(

**Comment by diegohaz** [comment]
Countless were the bugs I caught on https://github.com/reakit/reakit that were related to this issue. I use React Portal a lot and I can't think of a single case where I wanted event bubbling from portals to their parent components.

My workaround has been either checking it inside my parent event handlers:
```js
event.currentTarget.contains(event.target);
```
Or using native events instead:
```js
const onClick = () => {};
React.useEffect(() => {
  ref.current.addEventListener("click", onClick);
  return () => ref.current.removeEventListener("click", onClick);
});
```
I use those approaches internally in the library. But none of them are ideal. And, since this is an open source component library, I can't control how people pass event handlers to the components.

An option to disable event bubbling would solve all those problems.

**Comment by mogelbrod** [comment]
I hacked together a [semi-workaround](https://gist.github.com/mogelbrod/7b786f05297e55ea91169d2c9b99d55b) which blocks React bubbling while also retriggering a clone of the event on `window`. It appears to work in Chrome, Firefox and Safari on OSX, but IE11 is left out due to not allowing `event.target` to be manually set. So far it only cares about Mouse, Pointer, Keyboard and Wheel events. Not sure if drag events would be possible to clone.

Unfortunately it's not usable in our codebase since we require IE11 support, but maybe someone else could adapt it for their own use.

**Comment by Rijk** [comment]
What makes this especially mind boggling, is that the 'default' behaviour bubbles _down_ the component tree again. Take the following tree:

```jsx
<Link>
   <Menu (portal)>
      <form onSubmit={...}>
         <button type="submit">
```

I have been pulling my hair out for hours, as to why with this exact combination of components my form’s `onSubmit` is never called — regardless if I click the submit button or press Enter in an input field inside the form.

Finally, I discovered it is because the React Router [`Link`](https://reacttraining.com/react-router/web/api/Link) component has an `onClick` implementation that does `e.preventDefault()` to prevent the browser from reloading. However, this has the unfortunate side effect of also blocking the default behaviour of the click on the submit button, which happens to be submitting the form. So what I learned today is the onSubmit is actually called by the browser, as a default action to pressing the submit button. Even when you press enter, it triggers a click on the submit button, thus triggering a form submit.

But you see how the event bubbling order makes this really really weird.

1. `<input>` [keypress enter] 
2. `<button type="submit">` [simulated click] 
3. `<Menu>` [event propagates outside portal]
4. `<Link>` [propagation reaches parent `Link`] 
5. `<Link>` [calls `e.preventDefault()`] 
6. => default browser response to submit button click is canceled, event though we already passed that component, and `Link` has nothing to do with it and also didn’t intend to block this behaviour at all.

The solution for me (if anyone runs into the same problem) was the commonly used solution of wrapping the `<Menu>` content in a div with `onClick={e => e.stopPropagation()}`. But my point is I lost a lot of time tracking down the issue because the behaviour is really unintuitive.

**Comment by dcporter** [comment]
> The solution for me (if anyone runs into the same problem) was the commonly used solution of wrapping the `<Menu>` content in a div with `onClick={e => e.stopPropagation()}`. But my point is I lost a lot of time tracking down the issue because the behaviour is really unintuitive.

Yup — each _individual instance of the problem_ has the same easy solution, _once you've experienced the bug and correctly identified it_. It's a very steep-walled pit of failure that the React team has carved out here, and it's frustrating to not hear any acknowledgement of that from them.

**Comment by craigkovatch** [comment]
I've spent several days trying to debug another issue with mouseenter/mouseout bubbling out of portals unexpectedly. Even with `onMouseEnter={e => e.stopPropagation()}` on the portal'ed div, the events are still bubbling out to the owning Button, as in https://github.com/facebook/react/issues/11387#issuecomment-340009465 (the first comment on this issue). `mouseenter`/`mouseleave` aren't supposed to bubble in the first place...

 @gaearon @trueadm this issue has caused consistent, enormous frustration for over two years now. This thread is one of the top active issues on React. **Please**, could someone from the team contribute here?

**Comment by ProteanCode** [comment]
In my case opening the Window component by clicking on a Button made the window dissapear since the clicking on Window caused a click on button which caused the state change

I am new with React, I mostly use a jQuery and vanillia JS but this is mind blowing bug. There may be like 1% of cases when this behaviour would be expected...

**Comment by zbrogz** [comment]
I like the [two solutions](https://github.com/facebook/react/issues/11387#issuecomment-524113945) from @diegohaz, but I still think `createPortal` should have an option to stop event bubbling.

My particular use case was with a tooltip's `onMouseLeave` and `onMouseEnter` handlers being triggered by its child's portal descendant -- which wasn't desired. Native events fixed this by ignoring the portal descendants since they aren't dom descendants.

**Comment by delashum** [comment]
+1 for the option to stop bubbling in portals. It was suggested to just place the portal as a sibling (instead of a child) to the component where the event listener is originating, but i think that doesn't work in many use cases (including mine).

**Comment by mogelbrod** [comment]
It finally looks like `ReactDOM.unstable_renderSubtreeIntoContainer` [will be removed](https://github.com/facebook/react/pull/17880), meaning there soon won't be any reasonable workarounds left for this issue...

**Comment by craigkovatch** [comment]
^ help us @trueadm -nobi you're our only hope

**Comment by mogelbrod** [comment]
It looks like pinging them on GitHub doesn't work 😞
Perhaps somebody with an active Twitter account could tweet about this, tagging one of the contributors?

**Comment by justjake** [comment]
Adding my +1 for this issue. At Notion, we currently use a custom portal implementation that predates `React.createPortal`, and we manually forward our context providers to the new tree. I tried to adopt `React.createPortal` but was blocked by the unexpected bubbling behavior:

@sebmarkbage's suggestion to move the `<Portal>` outside of the `<MenuItem>` component to become a sibling only solves the issue for a single nesting level. The issue remains if you have multiple nested (eg) menu items that portal out sub-menus.

**Comment by justjake** [comment]
Bump.

**Comment by mogelbrod** [comment]
Dan left [a comment](https://github.com/facebook/react/issues/16721#issuecomment-674573888) on a related issue:
> @mogelbrod I don't currently have anything to add to that, but something like this ([#11387 (comment)](https://github.com/facebook/react/issues/11387#issuecomment-355395803)) seems reasonable to me if you're migrating an existing component.

**Comment by mogelbrod** [comment]
[Followup by Dan in the same issue](https://github.com/facebook/react/issues/16721#issuecomment-674807919):
> Thanks for the context about workaround. Since you already have that domain knowledge, the best next step is probably to write an RFC for the behavior you want and the alternatives you considered: https://github.com/reactjs/rfcs. Keep in mind that an RFC saying "let's just change this" is unlikely to be helpful. Writing a good RFC requires both understanding of why we have the current behavior, _and_ a plan to change it in a way that suits your use cases without regressing on others.
> 
> **Regardless of that**, `unstable_renderSubtreeIntoContainer` is not supported, so let's untangle these two discussions. We will not be adding propagation of Context to it because the whole API is frozen and will not be updated.

We should definitely publish a React RFC to suggest the addition of the discussed flag, or perhaps another solution. Does anybody feel particularly interested in drafting one (@justjake, @craigkovatch, or @jquense perhaps)? If not I'll see what I can come up with!

**Comment by jquense** [comment]
While I am interesting in evolving this API I don't have interest in drafting a RFC. Mostly bc it's a bunch of work and there is almost no chance it would be accepted, I don't think the core team actually considers RFCs that aren't already on their roadmap.

**Comment by gaearon** [comment]
@jquense I don't think this is accurate. Yes, we're unlikely to merge an RFC that doesn't align with the vision since adding a new API is always cross-cutting and influences all the other planned features. And it's fair we don't often comment on the ones that don't work. **However,** we do read through them, and especially when we approach a topic that the ecosystem has more expertise in. As examples, https://github.com/reactjs/rfcs/pull/38, https://github.com/reactjs/rfcs/pull/150, https://github.com/reactjs/rfcs/pull/118, https://github.com/reactjs/rfcs/pull/109, https://github.com/reactjs/rfcs/pull/32 have all been influential in our thinking even though we haven't explicitly commented on them.

**Comment by gaearon** [comment]
In other words, we approach RFCs in part as a community research mechanism. This comment from @mogelbrod (https://github.com/facebook/react/issues/16721#issuecomment-674748100) about why the workaround is annoying is exactly the kind of thing we'd like to see in an RFC. A consideration of existing solutions and their drawbacks can be more valuable than a concrete API suggestion proposal.

**Comment by jquense** [comment]
@gaearon My comment is not to suggest that the team does not listen to outside feedback. Ya'll do a good job of that. I do think my comment is accurate tho. The _process_ as it plays out on the RFC repo does not result in accepted RFC's from other folks. Looking at which RFCs are merged, it's entirely core team members or fb employees and no one else. Features that do make, ultimately are different and do not participate  the RFC process at all (e.g. isomorphic ID's).

I'm very glad to hear ya'll do look at the other RFC's and that they contribute into design for features, but "We were influenced by these outside RFC's even though we never commented on them", I think illustrates my point, not challenge it.

> In other words, we approach RFCs in part as a community research mechanism.

That's super reasonable, but not what the RFC repo says _its_ approach is, and not how other people generally think of RFC's. The RFC process are usually link and point of communication between the team and community, as well as something of an even playing field in terms of feature consideration and process. 

Larger points about community governance aside. Asking folks, to spend time writing up detailed proposals and then defend them to other outside participants while being met with silence from the react team is disheartening and actively strengthens the impression that FB only cares about it's own needs for OSS. Which sticks because i know ya'll don't feel  or design that way.

If the RFC process supposed to be: "here is where you can outline your concerns and use cases and we will read them, when/if we reach a point of being able to implement this feature". Honestly, that's a fine approach. I think the community would benefit from that that being explicitly spelt out, otherwise ppl will (and do) assume the same level of involvement and participation that other RFC processes often have and then be actively discouraged when that doesn't play out. I certainly have that impression even with slightly more insight than other contributors.

**Comment by gaearon** [comment]
Sure, I think I agree with all of that. I don't want to turn this into a meta-thread but just saying that since people keep pinging about this thread, *the most actionable thing to do* to move this forward is to write a proposal for how it should work that *takes concerns on both sides into account and compiles a deep understanding of the problem space*. I totally understand if this is not something people would like to sink their teeth in (in part based on how we respond to RFCs), which is why I have not suggested it earlier — but since I keep getting personal pings I wanted to suggest that as an option for someone who's motivated.

**Comment by jquense** [comment]
fair enough, this isn't the right place to get meta on RFCs :)

**Comment by craigkovatch** [comment]
@gaearon this is the #6 most-upvoted Issue currently open on React, and the #4 most commented. It has been open since React 16 was released, and is just 2 months away from being 3 years old now. Yet there has been very little engagement from the React core team. It feels very dismissive to say "it's up to the community to propose a solution" after that much time and pain has occurred. Please recognize that, though it does have some very useful applications, this behavior being the default was a design mistake. It should not be up to the community to RFC to fix it.

**Comment by gaearon** [comment]
I regret commenting on this issue and I retract my suggestion about the community RFC. You're right it's probably a bad idea. I have to add that this issue has become very emotionally charged, and as a human being I personally find it difficult to engage with it — even though I understand it is important and a lot of people feel strongly about it.

**Comment by gaearon** [comment]
Let me briefly reply on the state of this thread.

First, I want to apologize to the people who commented and have been frustrated by us not continuing the follow-ups in this thread. If I were reading this issue from the outside, my impression would probably have been that the React team has made a mistake, is unwilling to admit it, and are willing to sit on a simple solution ("just add one boolean, how hard can it be!") for over two years because they don't care about the community. I can totally understand how you might come to this conclusion.

I know this issue is highly upvoted. This has been brought up multiple times in this thread, maybe from the perspective that if React team had known this is a big pain point, we would have addressed it sooner. We *know* this is a pain point — people regularly message us privately about it, or hold it up as an example of how the React team doesn't care about the community. While I fully acknowledge that the silence has been frustrating, the mounting pressure to "just do something" has made it more difficult to engage with this issue productively.

This issue has workarounds — which makes it unlike a security vulnerability or a crash that has to be dealt with urgently. We know that wokarounds work (but aren't ideal and can be annoying) because we use some of them ourselves, especially around code that was written before React 16. I think we can agree that while this issue has undobtedly been frustrating to a large number of people, it is still in a different class of problems than a crash or a security issue that must be responded to within a concrete timeframe.

Additionally, I disagree with the framing that there is a simple solution we can implement tomorrow. *Even* if we consider the initial behavior a mistake (which I'm not sure I agree with), it means that the importance on making the next behavior correct is even *more* strict. If we fix some cases and break other cases, we haven't made any progress, and have created a ton of churn.

To give you an example, the current behavior is actually really helpful for the declaratively focus management use case that we have been researching for quite some time. It is useful to treat focus/blur as happening "inside" a modal with regards to the part tree, despite it being a Portal. If we were to ship the "simple" `createPortal(tree, boolean)` proposal suggested in this thread, this use case would not work because *the portal itself* cannot "know" about which behavior we want. Any exploration into a possible solution needs to consider dozens of use cases and some of them are not even fully understood yet. This is necessary to do at some point for sure, but it is also a huge time commitment to do it right, and so far we haven't been able to focus on it.

Generally, we as a team try to focus on few problems deeply, rather than on many problems shallowly. Unfortunately, this does mean that some conceptual flaws and gaps may not get filled for years because we are in the middle of fixing other important gaps or don't have an alternative design worked out that would have solved the problem for good. I know this is frustrating to hear, and it's a part of why I stayed away from this thread. Some other similar threads have turned into deeper explanations of the problems and possible solutions, which are helpful, but this one has mostly turned into a flood of "+1" and suggestions for a "simple" fix, which is why it has been difficult to engage with it meaningfully.

I know this isn't the answer people wanted to hear, but I hope it is better than no answer at all.

**Comment by gaearon** [comment]
Another thing that's worth calling out is that some of the pain points described in this thread might have been solved by other means. For example:

>Specifically: calling stopPropagation on the synthetic focus event to prevent it from bubbling out of the portal causes stopPropagation to also be called on the native focus event in React's captured handler on #document, which meant it did not make it to another captured handler on <body>

React no longer uses capture phase for emulating bubbling, and also doesn't listen to events on the document anymore. So without dismissing the frustration, it will definitely be necessary to reevaluate everything that has been posted so far in the light on the other changes.

**Comment by diegohaz** [comment]
@gaearon 

> Regarding the points about `renderSubtreeIntoContainer` being removed. Literally its only difference from `ReactDOM.render` is that it propagates Legacy Context. Since any release that wouldn't include `renderSubtreeIntoContainer` would also not include Legacy Context, `ReactDOM.render` would stay a 100% identical alternative. This doesn't, of course, solve the broader issue, but I think the concern with `renderSubtree` specifically is somewhat misplaced.

Now that you mentioned it, I wonder if the code below would be a valid and safe implementation for a React Portal without event bubbling:

```js
function Portal({ children }) {
  const containerRef = React.useRef();

  React.useEffect(() => {
    const container = document.createElement("div");
    containerRef.current = container;
    document.body.appendChild(container);
    return () => {
      ReactDOM.unmountComponentAtNode(container);
      document.body.removeChild(container);
    };
  }, []);

  React.useEffect(() => {
    ReactDOM.render(children, containerRef.current);
  }, [children]);

  return null;
}
```

CodeSandbox with some tests: https://codesandbox.io/s/react-portal-with-reactdom-render-m22dj?file=/src/App.js:0-538

**Comment by gaearon** [comment]
There is still an issue with not passing Modern Context through but this isn't a new issue (`renderSubtree` is also affected by it). The workaround there is to surround your tree with a bunch of context providers. Overall it's not ideal to nest trees, so I wouldn't recommend to shift way to this pattern in anything other than legacy existing code scenarios.

**Comment by mogelbrod** [comment]
Once again, thank you so much for the writeup @gaearon!

It sounds like aggregating the list of broken cases + workarounds (updated for React v17) would be the most productive thing for somebody outside the core team (correct me if I'm wrong!).

I'm swamped the coming weeks but aim to do ASAP. If anybody else is able to do this earlier, or chime in with snippets (like @diegohaz just did), that would be awesome!

**Comment by gaearon** [comment]
Aggregating a list of cases would definitely be useful, although I'd say it needs to include not only broken cases but also the cases where the current behavior makes sense.

**Comment by jquense** [comment]
If there is a public space to add to i'd be happy to add use cases from both our apps and as a UI library author. In general i agree with dan that while sometimes annoying its easy to work around, and for cases where you do want bubbling it's very hard to do cover the case without React's help.

**Comment by mogelbrod** [comment]
> Aggregating a list of cases would definitely be useful, although I'd say it needs to include not only broken cases but also the cases where the current behavior makes sense.

I'd be happy to do include these if anybody can point me to some open source code/extracted code that relies on it! Like you mentioned previously it's a bit of a challenge to find since only people having issues with the current behaviour are involved on this issue 😅 

> If there is a public space to add to i'd be happy to add use cases from both our apps and as a UI library author. In general i agree with Dan that while sometimes annoying it's easy to work around. For cases where you do want React's bubbling it's very hard to do cover the case without React's help.

Any specific space you have in mind, or would sharing one [codesandbox](https://codesandbox.io) (or jsfiddle, etc.) per case work as a starter? I can try compiling them all once we've gathered some cases.

**Comment by gaearon** [comment]
I started a thread here: https://github.com/facebook/react/issues/19637.

**Comment by dcporter** [comment]
Bump!

**Comment by MatthieuCoelho** [comment]
It seems like onSubmit isn't propagated

**Comment by AndyOGo** [comment]
@gaearon 
I think the obvious solution is to make event bubbling and context sharing configurable.

The only question to me is how granular should this config be?
- just a boolean flag, like:
   - `bubbleEvents`
   - and `shareContext`
- or a list of bubbled events, like:
   - `bubbledEvent: ["click", "keydown", ...]`
   - `sharedContexts: [MyContext, ...]`

**Comment by haydar9** [comment]
> 
> 
> The workaround that worked for me is calling `stopPropagation` directly under portal rendering:
> 
> ```
> return createPortal(
>       <div onClick={e => e.stopPropagation()}>{this.props.children}</div>,
>       this.el
>     )
> ```
> 
> That works great for me since I have single abstraction component that uses portals, otherwise you will need to fix up all your `createPortal` calls.

> 
> 
> The workaround that worked for me is calling `stopPropagation` directly under portal rendering:
> 
> ```
> return createPortal(
>       <div onClick={e => e.stopPropagation()}>{this.props.children}</div>,
>       this.el
>     )
> ```
> 
> That works great for me since I have single abstraction component that uses portals, otherwise you will need to fix up all your `createPortal` calls.

This worked for me.

**Comment by nibmz7** [comment]
Hi. I just published a library and it contains a wrapper around Portals and it dodges this event bubbling behaviour by portal-ling the portal itself. I didn't really test the implementation but feel free to check it out. 

![image](https://user-images.githubusercontent.com/14259666/121818551-70992480-ccba-11eb-83d7-e8c1f539dd3c.png)

Portal Renderer code - https://github.com/nibmz7/nimz-app/blob/main/core/utils/Portal/PortalRenderer.tsx
Portal code - https://github.com/nibmz7/nimz-app/blob/main/core/utils/Portal/index.tsx
The library if you intend to try it out - https://www.npmjs.com/package/nimz-utils

I haven't tried the library out myself actually 😅. If it doesn't work, you may copy for portals over to your project.

**Comment by justjake** [comment]
Bump.

**Comment by dcporter** [comment]
Bump!

**Comment by pie6k** [comment]
My use case is quite extreme and exotic, but I'd really want to disable react portal event bubbling:

Use case:
- electron App
- we have multiple windows
- new window is created with `window.open()`
- we create react portal to this new window
- we render content from main window to this new window via portal
- event happening in new window gets bubbled to main window, which is totally exotic, as window gets events that happened in different window

**Comment by deniskabana** [comment]
Bump... I have a few modal windows I am managing from inside a huge container that has a lot of mouse events (drag, click, etc.) that I know nothing of at runtime. Can not stop event bubbling on unknown/all events...

**Comment by princ3od** [comment]
The issue still appears today. I can not stop propagation `mouseeenter` event from children to their parent.

**Comment by pie6k** [comment]
I agree. Portals IMO should have some sort of bubblingMode option with 3 modes: follow Dom, follow react, don't follow.

I use React portal to render stuff into another window and bubbling events in this case is really not needed/expected

**Comment by craigkovatch** [comment]
[Bubbling across portal would have been much better](https://xkcd.com/567/)

**Comment by devongovett** [comment]
We ran across this issue yet again in React Aria. Seriously considering adding this to all of our overlay components to hopefully solve it once and for all. 😐

```js
function Portal({children, container}) {
  let element = (
    <div
      onCopy={e => e.stopPropagation()}
      onCopyCapture={e => e.stopPropagation()}
      onCut={e => e.stopPropagation()}
      onCutCapture={e => e.stopPropagation()}
      onPaste={e => e.stopPropagation()}
      onPasteCapture={e => e.stopPropagation()}
      onCompositionEnd={e => e.stopPropagation()}
      onCompositionEndCapture={e => e.stopPropagation()}
      onCompositionStart={e => e.stopPropagation()}
      onCompositionStartCapture={e => e.stopPropagation()}
      onCompositionUpdate={e => e.stopPropagation()}
      onCompositionUpdateCapture={e => e.stopPropagation()}
      onFocus={e => e.stopPropagation()}
      onFocusCapture={e => e.stopPropagation()}
      onBlur={e => e.stopPropagation()}
      onBlurCapture={e => e.stopPropagation()}
      onChange={e => e.stopPropagation()}
      onChangeCapture={e => e.stopPropagation()}
      onBeforeInput={e => e.stopPropagation()}
      onBeforeInputCapture={e => e.stopPropagation()}
      onInput={e => e.stopPropagation()}
      onInputCapture={e => e.stopPropagation()}
      onReset={e => e.stopPropagation()}
      onResetCapture={e => e.stopPropagation()}
      onSubmit={e => e.stopPropagation()}
      onSubmitCapture={e => e.stopPropagation()}
      onInvalid={e => e.stopPropagation()}
      onInvalidCapture={e => e.stopPropagation()}
      onLoad={e => e.stopPropagation()}
      onLoadCapture={e => e.stopPropagation()}
      onError={e => e.stopPropagation()}
      onErrorCapture={e => e.stopPropagation()}
      onKeyDown={e => e.stopPropagation()}
      onKeyDownCapture={e => e.stopPropagation()}
      onKeyPress={e => e.stopPropagation()}
      onKeyPressCapture={e => e.stopPropagation()}
      onKeyUp={e => e.stopPropagation()}
      onKeyUpCapture={e => e.stopPropagation()}
      onAbort={e => e.stopPropagation()}
      onAbortCapture={e => e.stopPropagation()}
      onCanPlay={e => e.stopPropagation()}
      onCanPlayCapture={e => e.stopPropagation()}
      onCanPlayThrough={e => e.stopPropagation()}
      onCanPlayThroughCapture={e => e.stopPropagation()}
      onDurationChange={e => e.stopPropagation()}
      onDurationChangeCapture={e => e.stopPropagation()}
      onEmptied={e => e.stopPropagation()}
      onEmptiedCapture={e => e.stopPropagation()}
      onEncrypted={e => e.stopPropagation()}
      onEncryptedCapture={e => e.stopPropagation()}
      onEnded={e => e.stopPropagation()}
      onEndedCapture={e => e.stopPropagation()}
      onLoadedData={e => e.stopPropagation()}
      onLoadedDataCapture={e => e.stopPropagation()}
      onLoadedMetadata={e => e.stopPropagation()}
      onLoadedMetadataCapture={e => e.stopPropagation()}
      onLoadStart={e => e.stopPropagation()}
      onLoadStartCapture={e => e.stopPropagation()}
      onPause={e => e.stopPropagation()}
      onPauseCapture={e => e.stopPropagation()}
      onPlay={e => e.stopPropagation()}
      onPlayCapture={e => e.stopPropagation()}
      onPlaying={e => e.stopPropagation()}
      onPlayingCapture={e => e.stopPropagation()}
      onProgress={e => e.stopPropagation()}
      onProgressCapture={e => e.stopPropagation()}
      onRateChange={e => e.stopPropagation()}
      onRateChangeCapture={e => e.stopPropagation()}
      onSeeked={e => e.stopPropagation()}
      onSeekedCapture={e => e.stopPropagation()}
      onSeeking={e => e.stopPropagation()}
      onSeekingCapture={e => e.stopPropagation()}
      onStalled={e => e.stopPropagation()}
      onStalledCapture={e => e.stopPropagation()}
      onSuspend={e => e.stopPropagation()}
      onSuspendCapture={e => e.stopPropagation()}
      onTimeUpdate={e => e.stopPropagation()}
      onTimeUpdateCapture={e => e.stopPropagation()}
      onVolumeChange={e => e.stopPropagation()}
      onVolumeChangeCapture={e => e.stopPropagation()}
      onWaiting={e => e.stopPropagation()}
      onWaitingCapture={e => e.stopPropagation()}
      onAuxClick={e => e.stopPropagation()}
      onAuxClickCapture={e => e.stopPropagation()}
      onClick={e => e.stopPropagation()}
      onClickCapture={e => e.stopPropagation()}
      onContextMenu={e => e.stopPropagation()}
      onContextMenuCapture={e => e.stopPropagation()}
      onDoubleClick={e => e.stopPropagation()}
      onDoubleClickCapture={e => e.stopPropagation()}
      onDrag={e => e.stopPropagation()}
      onDragCapture={e => e.stopPropagation()}
      onDragEnd={e => e.stopPropagation()}
      onDragEndCapture={e => e.stopPropagation()}
      onDragEnter={e => e.stopPropagation()}
      onDragEnterCapture={e => e.stopPropagation()}
      onDragExit={e => e.stopPropagation()}
      onDragExitCapture={e => e.stopPropagation()}
      onDragLeave={e => e.stopPropagation()}
      onDragLeaveCapture={e => e.stopPropagation()}
      onDragOver={e => e.stopPropagation()}
      onDragOverCapture={e => e.stopPropagation()}
      onDragStart={e => e.stopPropagation()}
      onDragStartCapture={e => e.stopPropagation()}
      onDrop={e => e.stopPropagation()}
      onDropCapture={e => e.stopPropagation()}
      onMouseDown={e => e.stopPropagation()}
      onMouseDownCapture={e => e.stopPropagation()}
      onMouseEnter={e => e.stopPropagation()}
      onMouseLeave={e => e.stopPropagation()}
      onMouseMove={e => e.stopPropagation()}
      onMouseMoveCapture={e => e.stopPropagation()}
      onMouseOut={e => e.stopPropagation()}
      onMouseOutCapture={e => e.stopPropagation()}
      onMouseOver={e => e.stopPropagation()}
      onMouseOverCapture={e => e.stopPropagation()}
      onMouseUp={e => e.stopPropagation()}
      onMouseUpCapture={e => e.stopPropagation()}
      onSelect={e => e.stopPropagation()}
      onSelectCapture={e => e.stopPropagation()}
      onTouchCancel={e => e.stopPropagation()}
      onTouchCancelCapture={e => e.stopPropagation()}
      onTouchEnd={e => e.stopPropagation()}
      onTouchEndCapture={e => e.stopPropagation()}
      onTouchMove={e => e.stopPropagation()}
      onTouchMoveCapture={e => e.stopPropagation()}
      onTouchStart={e => e.stopPropagation()}
      onTouchStartCapture={e => e.stopPropagation()}
      onPointerDown={e => e.stopPropagation()}
      onPointerDownCapture={e => e.stopPropagation()}
      onPointerMove={e => e.stopPropagation()}
      onPointerMoveCapture={e => e.stopPropagation()}
      onPointerUp={e => e.stopPropagation()}
      onPointerUpCapture={e => e.stopPropagation()}
      onPointerCancel={e => e.stopPropagation()}
      onPointerCancelCapture={e => e.stopPropagation()}
      onPointerEnter={e => e.stopPropagation()}
      onPointerEnterCapture={e => e.stopPropagation()}
      onPointerLeave={e => e.stopPropagation()}
      onPointerLeaveCapture={e => e.stopPropagation()}
      onPointerOver={e => e.stopPropagation()}
      onPointerOverCapture={e => e.stopPropagation()}
      onPointerOut={e => e.stopPropagation()}
      onPointerOutCapture={e => e.stopPropagation()}
      onGotPointerCapture={e => e.stopPropagation()}
      onGotPointerCaptureCapture={e => e.stopPropagation()}
      onLostPointerCapture={e => e.stopPropagation()}
      onLostPointerCaptureCapture={e => e.stopPropagation()}
      onScroll={e => e.stopPropagation()}
      onScrollCapture={e => e.stopPropagation()}
      onWheel={e => e.stopPropagation()}
      onWheelCapture={e => e.stopPropagation()}
      onAnimationStart={e => e.stopPropagation()}
      onAnimationStartCapture={e => e.stopPropagation()}
      onAnimationEnd={e => e.stopPropagation()}
      onAnimationEndCapture={e => e.stopPropagation()}
      onAnimationIteration={e => e.stopPropagation()}
      onAnimationIterationCapture={e => e.stopPropagation()}
      onTransitionEnd={e => e.stopPropagation()}
      onTransitionEndCapture={e => e.stopPropagation()}
    >
      {children}
    </div>
  );

  return ReactDOM.createPortal(element, container);
}
```

**Comment by craigkovatch** [comment]
I'm currently working on a bug from five years ago that goes all the way back to this issue. Really sad this has been here since Oct 2017 with no official response at all. I think this is the worst API decision decision in React and am confused why there's still no ...whatever the opposite of an escape hatch is

**Comment by zombieJ** [comment]
Hmmm...
I could not remember how many times I track and found this issue...

**Comment by YassienW** [comment]
Stopping propagation is NOT a solution, sometimes you want to handle events further down the line or on the document. It makes perfect sense to be able to propagate events according to their natural location in the DOM (i'd argue even more sense than what we have now)

Like all others in this thread I'm now facing an issue where nested elements that use portals propagate events and affect each other in weird illogical ways, with no straight forward way to fix it.

**Comment by vrajpal-jhala** [comment]
While `.stopPropagation` do work most of the time,  there are cases when you want bubble it for only the document and skipping every other parent. ie. A modal with mouse events for drag and drop. We ended up dispatching a cloned event specifically for the document (global events).

**Comment by AbdelrahmanAbounida** [comment]
I solved it like that 




```

{/** child **}
<div
        onMouseDown={(event) => {
          event.stopPropagation();
        }}
        className="p-4 font-semibold border-b-2 text-left flex flex-row cursor-default justify-between items-center"
      >
You can't  drag me 
</div> 

```

**Comment by ElSalvo96** [comment]
Hello, I encountered the same issue with this version in my company's UI Library:
```
    "react": "18.2.0"
    "react-dom": "18.2.0"
```

I also found an easy way to run into this issue. 
Just go there https://react-bootstrap.netlify.app/docs/components/modal/#live-demo and edit the component from:
```
function Example() {
  const [show, setShow] = useState(false);

  const handleClose = () => setShow(false);
  const handleShow = () => setShow(true);

  return (
    <>
      <Button variant="primary" onClick={handleShow}>
        Launch demo modal
      </Button>

      <Modal show={show} onHide={handleClose}>
        <Modal.Header closeButton>
          <Modal.Title>Modal heading</Modal.Title>
        </Modal.Header>
        <Modal.Body>Woohoo, you are reading this text in a modal!</Modal.Body>
        <Modal.Footer>
          <Button variant="secondary" onClick={handleClose}>
            Close
          </Button>
          <Button variant="primary" onClick={handleClose}>
            Save Changes
          </Button>
        </Modal.Footer>
      </Modal>
    </>
  );
}
```

To:

```
function Example() {
  const [show, setShow] = useState(false);

  const handleClose = () => setShow(false);
  const handleShow = () => setShow(true);

  return (
    <>
      <Button variant="primary" onClick={handleShow}>
        Launch demo modal
        <Modal show={show} onHide={handleClose}>
          <Modal.Header closeButton>
            <Modal.Title>Modal heading</Modal.Title>
          </Modal.Header>
          <Modal.Body>Woohoo, you are reading this text in a modal!</Modal.Body>
          <Modal.Footer>
            <Button variant="secondary" onClick={handleClose}>
              Close
            </Button>
            <Button variant="primary" onClick={handleClose}>
              Save Changes
            </Button>
          </Modal.Footer>
        </Modal>
      </Button>
    </>
  );
}
```

With the second component you can exit using only esc button. Of course, this is not a real-world example, but when you have layered components in a large codebase, you can easily run into this issue.

**Comment by tobq** [comment]
Insane to me that this hasn't been resolved after 7 years of the community complaining. Clearly a bad initial API design. `stopPropagation` is not a real solution - and only leads to finicky bugs later down the line.

**Comment by rjgotten** [comment]
@jasosims 
> If you want things like dropdowns and popovers to close when clicking outside of them, the only reasonable approach is to add an event listener to `document.body`. If something receives a click and stops propagation, it won't work.

If you're just dealing with things like custom styled selects, aka `role="combobox"` from ARIA, then there's a better way.
You keep the native browser focus on the opener element and use virtual focus with `aria-activedescendant` to track focused items within the list. And you close the dropdown popover when the opener loses focus.

For those cases where you need to actually be able to tap on multiple elements before a popover closes (e.g. if you make a dropdown list `aria-multiselectable="true"`) you can make it work with a fairly simple recipe, as long as you have a ref to the popup DOM element and use native event handlers. (This doesn't work with synthetic events.)

1. set `touchAction = "pan"` on the list element to ensure scrolling gestures will cancel touches and pointer events.
2. attach a `mousedown` listener that calls `event.preventDefault()`
3. attach a `touchend` listener that calls `event.preventDefault()`
4. attach a `pointerup` listener that checks whether `event.isPrimary` and `event.pointerType == "touch"` and if so, use the pointer event to create a new `PointerEvent` of type `"click"` and dispatch it off of `event.target`.

This blocks focus moving off of the opener element, but _does_ still give you a working click event on whatever is inside your popover that you can then work with. It's a portable recipe that can easily be encapsulated into a custom hook.

**Comment by krutoo** [comment]
This terrible feature of React portals is annoying every time you encounter it

I fully support that we need an option to disable this behavior.

**Comment by gkurt** [comment]
I always assumed this was a bug with the component libraries I use. TIL it's a React issue.

I like React very much when it doesn't enforce opinionated stuff. But this is quite bad. The only reparenting feature in React has this issue, without a way to disable it. Do I have to implement a custom portal now too?
